File: prolog/check_axis_boundary.pl
% ============================================================================
% CHECK AXIS BOUNDARY — transitive committer→observer taint guard (OQ-15 / OQ-135)
% ============================================================================
% Implements v8 §8 item 1 (LOAD-BEARING): the one-seat invariant as a
% reachability / dataflow property over the *actual loaded call graph*, NOT a
% count or an import grep (v8 calls those a trap; W1 confirmed grep is blind —
% cs_drift_mismatch reaches observer machinery transitively, audit
% 2026-06-23_oq15_crossaxis_witnesses/).
%
% THE INVARIANT (v8 §3, directional):
%   No committer field reaches OBSERVER computation by ANY path except as the
%   entailment-typed payload on the single forward bridge
%   `influences` → detect_necessity_inheritance.
%
% HOW THIS REALIZES IT.  The committer axis is uniformly `cs_*`-prefixed (the
% facts cs_reading_relation/cs_kernel_id/cs_story_uid/cs_axiom*/cs_drift_state
% and every derived cs_* predicate). A "committer field reaching observer
% computation" is therefore a call edge  P → cs_X  in a loaded clause body where
% P is NOT itself committer-side (`cs_*`) and NOT in the mediator-output layer
% (json_report — the sanctioned both-axis reader that writes only JSON). Any
% transitive observer→committer path crosses the boundary at exactly one such
% last-hop edge (the last non-cs_ caller before the first cs_ callee), so the
% set of these DIRECT edges is COMPLETE for the guarded direction — transitive
% reachability collapses to edge-enumeration over the call graph. This reads the
% loaded program (clause/2 over every engine-defined predicate's bodies),
% descending into control constructs and meta-calls; it is not text matching.
%
% WHAT IT EMITS.  One `AXIS_EDGE: <SrcMod>:<Name>/<Ar> -> <Callee>/<Ar>` line per
% boundary-crossing edge, sorted+deduped. The Python harness
% (python/check_axis_boundary.py) diffs these against
% prolog/axis_boundary_allowlist.txt (load_warning_gate.py pattern): any edge not
% allowlisted → exit 1; an allowlisted edge that has disappeared → stale note.
%
% Run (engine only — corpus not needed; the call graph is static):
%   cd prolog && swipl -l check_axis_boundary.pl -g "run_axis_boundary, halt" -t "halt(1)"
% ============================================================================

:- [stack].

% Source files we count as ENGINE (own the call graph). File-based, so it is
% robust to module aliasing; excludes SWI library and this checker itself.
engine_file(F) :-
    sub_atom(F, _, _, _, '/prolog/'),
    \+ sub_atom(F, _, _, _, '/library/'),
    \+ sub_atom(F, _, _, _, 'check_axis_boundary').

% A predicate defined in the engine, with its DEFINING module + file. The
% imported_from/reexport guard is essential: drl_modal_logic `:- reexport(
% drl_composition)` etc. would otherwise surface detect_necessity_inheritance
% under every importing module — a module-attribution artifact, not real edges.
engine_predicate(M, Name, Arity, Head, F) :-
    current_module(M),
    current_predicate(Name, M:Head),
    functor(Head, Name, Arity),
    \+ predicate_property(M:Head, imported_from(_)),   % defining module only
    catch(predicate_property(M:Head, file(F)), _, fail),
    engine_file(F),
    % only predicates with walkable clauses (skip foreign/built-in)
    \+ predicate_property(M:Head, foreign),
    \+ predicate_property(M:Head, built_in).

% MEDIATOR-OUTPUT layer: reads both axes by design, writes only to JSON. A
% committer read from here is NOT a violation (it is the mediator's job). Keep
% this list MINIMAL and explicit — it is a sanctioned exemption, not a default.
mediator_module(json_report).

% body_calls(+Body, -CalleeName/-CalleeArity): every goal called in a clause
% body, descending through control constructs and the common meta-call wrappers.
body_calls(V, _) :- var(V), !, fail.
body_calls((A,B), G)        :- !, ( body_calls(A,G) ; body_calls(B,G) ).
body_calls((A;B), G)        :- !, ( body_calls(A,G) ; body_calls(B,G) ).
body_calls((A->B), G)       :- !, ( body_calls(A,G) ; body_calls(B,G) ).
body_calls((A*->B), G)      :- !, ( body_calls(A,G) ; body_calls(B,G) ).
body_calls(\+(A), G)        :- !, body_calls(A,G).
body_calls(not(A), G)       :- !, body_calls(A,G).
body_calls(call(A), G)      :- !, body_calls(A,G).
body_calls(once(A), G)      :- !, body_calls(A,G).
body_calls(ignore(A), G)    :- !, body_calls(A,G).
body_calls(findall(_,A,_), G):- !, body_calls(A,G).
body_calls(findall(_,A,_,_), G):- !, body_calls(A,G).
body_calls(forall(A,B), G)  :- !, ( body_calls(A,G) ; body_calls(B,G) ).
body_calls(aggregate_all(_,A,_), G) :- !, body_calls(A,G).
body_calls(bagof(_,A,_), G) :- !, body_calls(A,G).
body_calls(setof(_,A,_), G) :- !, body_calls(A,G).
body_calls(catch(A,_,B), G) :- !, ( body_calls(A,G) ; body_calls(B,G) ).
body_calls(_^A, G)          :- !, body_calls(A,G).
body_calls(maplist(Goal,_), G)     :- !, meta_partial(Goal, G).
body_calls(maplist(Goal,_,_), G)   :- !, meta_partial(Goal, G).
body_calls(maplist(Goal,_,_,_), G) :- !, meta_partial(Goal, G).
% Peel module qualifiers RECURSIVELY: a clause consulted outside a module file
% can store its body double-qualified, e.g. user:(narrative_ontology:cs_X(..)).
% Taking functor/3 of the inner term then yields (:)/2 and the cs_ sink is missed
% (caught by positive control 1, 2026-06-23). Recurse to reach the real callee.
body_calls(M:Goal, G) :- !, atom(M), body_calls(Goal, G).
body_calls(Goal, N/A) :- callable(Goal), functor(Goal, N, A).

% maplist(Closure, ...): the closure names the called predicate.
meta_partial(Goal, N/A0) :-
    callable(Goal),
    ( Goal = M:G2, atom(M) -> true ; G2 = Goal ),
    functor(G2, N, A),
    A0 = A.   % under-counts added maplist args; name is what the cs_ test needs

% A committer taint sink: a predicate whose name is cs_-prefixed (covers the
% narrative_ontology committer facts AND every derived cs_* predicate).
committer_name(Name) :- atom(Name), sub_atom(Name, 0, 3, _, 'cs_').

% Source is committer-SIDE (not observer) if it is defined in a committer module
% (file basename cs_*) OR its own name is cs_-prefixed. Module-based catches the
% committer predicates that are NOT cs_-named (compare_kernel_readings in
% cs_kernel_registry, run_drift_mismatch_report in cs_drift_mismatch) — they read
% cs_ facts as their OWN axis, not across the boundary.
committer_side(_M, Name, _F) :- committer_name(Name), !.
committer_side(_M, _Name, F) :- file_base_name(F, B), sub_atom(B, 0, 3, _, 'cs_').

% boundary_edge(SrcMod, SrcName/SrcAr, CalleeName/CalleeAr): committer→observer
% crossing — an observer-side / tooling caller reaching a cs_ sink.
boundary_edge(M, Name/Arity, Callee) :-
    engine_predicate(M, Name, Arity, Head, F),
    \+ committer_side(M, Name, F),  % source is NOT committer-side
    \+ mediator_module(M),          % source is NOT the mediator-output layer
    catch(clause(M:Head, Body), _, fail),
    body_calls(Body, Callee),
    Callee = CName/_,
    committer_name(CName).          % sink IS committer-side

run_axis_boundary :-
    findall(edge(M, Src, Callee), boundary_edge(M, Src, Callee), Raw),
    sort(Raw, Edges),
    ( Edges == []
    ->  format("AXIS_EDGE_NONE~n", [])
    ;   forall(member(edge(M, Name/Ar, CName/CAr), Edges),
               format("AXIS_EDGE: ~w:~w/~w -> ~w/~w~n", [M, Name, Ar, CName, CAr]))
    ),
    length(Edges, N),
    format("AXIS_EDGE_COUNT: ~w~n", [N]).

File: python/check_axis_boundary.py
#!/usr/bin/env python3
"""check_axis_boundary.py — fail loud on UNEXPECTED committer→observer reads (OQ-15 / OQ-135).

The invariant this guards (v8 §8 item 1, LOAD-BEARING): no committer field
reaches observer computation by any path except as the entailment-typed payload
on the single forward bridge `influences` → detect_necessity_inheritance. This is
a *reachability* property, not a count or import grep (v8 calls those a trap; W1
confirmed grep is blind — cs_drift_mismatch reaches observer machinery
transitively, audit 2026-06-23_oq15_crossaxis_witnesses/).

The reachability core is prolog/check_axis_boundary.pl: it walks the LOADED call
graph (clause/2 over every engine-defined predicate's bodies, descending through
control constructs and meta-calls) and emits one `AXIS_EDGE:` line per
committer→observer boundary-crossing edge. Any transitive observer→committer path
crosses the boundary at exactly one such last-hop edge, so the direct-edge set is
complete for the guarded direction.

This harness mirrors load_warning_gate.py: collect the edges, diff against the
allowlist (prolog/axis_boundary_allowlist.txt), and:
  - any edge NOT allowlisted          -> exit 1, listed (a new cross-axis read)
  - allowlisted edge no longer present -> note printed (stale entry; exit 0)

Usage:
    python3 python/check_axis_boundary.py            # gate (exit 1 on unexpected)
    python3 python/check_axis_boundary.py --baseline # print current edge set
    python3 python/check_axis_boundary.py --allowlist PATH   # alternate allowlist
                                                     # (used by the positive controls)
"""
import argparse
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
PROLOG = ROOT / "prolog"
ALLOWLIST = PROLOG / "axis_boundary_allowlist.txt"
PROBE = "check_axis_boundary.pl"


def collect_edges(probe=PROBE):
    """Run the reachability probe; return the sorted set of boundary edges.

    `probe` lets the positive controls point at a wrapper that loads the real
    check_axis_boundary.pl plus a planted cross-axis read — same harness, same
    allowlist, so a planted violation must drive exit 1 end-to-end."""
    proc = subprocess.run(
        ["swipl", "-l", probe, "-g", "run_axis_boundary, halt", "-t", "halt(1)"],
        cwd=PROLOG, capture_output=True, text=True, timeout=300,
    )
    edges = []
    for line in proc.stdout.splitlines():
        line = line.strip()
        if line.startswith("AXIS_EDGE:"):
            edges.append(line[len("AXIS_EDGE:"):].strip())
    if proc.returncode != 0 and not edges:
        # the probe itself failed to run — fail closed, surface stderr
        print(proc.stderr, file=sys.stderr)
        print("[AXIS-GATE] reachability probe did not run cleanly", file=sys.stderr)
    return sorted(set(edges)), proc.returncode


def load_allowlist(path):
    allowed = set()
    p = Path(path)
    if p.exists():
        for ln in p.read_text().splitlines():
            ln = ln.strip()
            if ln and not ln.startswith("#"):
                allowed.add(ln)
    return allowed


def _gate(probe, allowlist=str(ALLOWLIST)):
    """Return (n_unexpected, exit_code) for one gate run — used by selftest."""
    edges, rc = collect_edges(probe)
    allowed = load_allowlist(allowlist)
    unexpected = [e for e in edges if e not in allowed]
    code = 1 if (rc != 0 and not edges) else (1 if unexpected else 0)
    return unexpected, code


def selftest():
    """Negative case passes; BOTH planted-violation controls fire. The clean
    pass is only trusted because these show the guard WOULD flag a violation
    (path b: count-blind payload widening; path c: per-bridge-blind seam)."""
    cases = [
        ("negative (clean corpus)",        PROBE,                            0),
        ("control1 payload-widen (path b)", "tests/axis_boundary_ctl_run1.pl", 1),
        ("control2 nonbridge-seam (path c)", "tests/axis_boundary_ctl_run2.pl", 1),
    ]
    ok = True
    for label, probe, want in cases:
        unexpected, code = _gate(probe)
        verdict = "PASS" if code == want else "FAIL"
        if code != want:
            ok = False
        detail = (" :: " + "; ".join(unexpected)) if unexpected else ""
        print(f"[AXIS-SELFTEST] {verdict}  {label}: exit {code} (want {want}){detail}")
    print(f"[AXIS-SELFTEST] {'ALL PASS' if ok else 'FAILURES PRESENT'}")
    return 0 if ok else 1


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--baseline", action="store_true",
                    help="print current boundary-edge set and exit 0")
    ap.add_argument("--allowlist", default=str(ALLOWLIST))
    ap.add_argument("--probe", default=PROBE,
                    help="alternate probe entry file (positive-control wrappers)")
    ap.add_argument("--selftest", action="store_true",
                    help="run negative case + both planted-violation positive controls")
    args = ap.parse_args()

    if args.selftest:
        return selftest()

    edges, rc = collect_edges(args.probe)
    if args.baseline:
        for e in edges:
            print(e)
        return 0

    allowed = load_allowlist(args.allowlist)
    unexpected = [e for e in edges if e not in allowed]
    stale = sorted(allowed - set(edges))

    for e in unexpected:
        print(f"[AXIS-GATE] UNEXPECTED committer→observer read: {e}", file=sys.stderr)
    for e in stale:
        print(f"[AXIS-GATE] stale allowlist entry (edge gone — prune it): {e}")
    print(f"[AXIS-GATE] {len(edges)} boundary edges: "
          f"{len(edges) - len(unexpected)} allowlisted, {len(unexpected)} unexpected")
    # fail closed if the probe could not run at all
    if rc != 0 and not edges:
        return 1
    return 1 if unexpected else 0


if __name__ == "__main__":
    sys.exit(main())
